Source/JavaScriptCore/wasm/WasmSectionParser.cpp

```diff
@@ -63,12 +64,23 @@ auto SectionParser::parseType() -> PartialResult
             WASM_FAIL_IF_HELPER_FAILS(parseStructType(i, signature));
             break;
         }
+        case TypeKind::Rec: {
+            if (!Options::useWebAssemblyGC())
+                return fail(i, "th type failed to parse because rec types are not enabled");
+
+            WASM_FAIL_IF_HELPER_FAILS(parseRecursionGroup(i, signature));
+            break;
+        }
         default:
-            return fail(i, "th Type is non-Func ", typeKind);
+            return fail(i, "th Type is non-Func and non-Struct ", typeKind);
         }
 
         WASM_PARSER_FAIL_IF(!signature, "can't allocate enough memory for Type section's ", i, "th signature");
-        m_info->typeSignatures.uncheckedAppend(signature.releaseNonNull());
+        // Recursion group parsing will append the entries itself, as there may
```

Comment:
Hmm, I wish these cases both acted the same way. 

Consider ```
(struct $a i32)
(rec (struct $b i32))
(struct $c i32)
```
Is it the case that $a != $b != $c, since they are all defined in separate recursion groups? In the parlance of the spec draft, they have different tied context types?

Reply:
Either way, it would be nice for all the cases to act the same way.

Reply:
These should actually all be equivalent, I think it's just an oversight/bug in the code. You can verify this by trying this example in the reference interpreter:

```
(module 
  (type $a (struct (field i32)))
  (rec (type $b (struct (field i32))))
  (type $c (struct (field i32)))

  (func (result (ref null $a)) (ref.null $b))
  (func (result (ref null $a)) (ref.null $c))
  (func (result (ref null $b)) (ref.null $a))
  (func (result (ref null $b)) (ref.null $c))
  (func (result (ref null $c)) (ref.null $a))
  (func (result (ref null $c)) (ref.null $b)))
```

which should validate AFAICT.

The issue in the code is that either we have to normalize structs without a recursion group to be represented with a singleton recursion group, or we have to normalize non-recursive, single-member recursion groups into just a struct definition.

I think the latter approach is probably better, as it avoids the need to unroll projections when they're not needed. I did something similar for normalizing subtyping declarations.

Reply:
I just fixed this case in the latest push, and added a test case for the above example. It normalizes non-recursive, single recursion groups as being just the underlying type definition.

Parsing still has multiple code paths for this though, because the binary shorthand may or may not be used.